### PR TITLE
Blocks: Try to make MediaUpload component extensible

### PR DESCRIPTION
**Pull request by gziolo**
## Description

At the moment `MediaUpload` component depends on `wp.media`, which makes also `blocks` module dependent on `media` module.  This PR tries to move this dependency to the `edit-post` in Gutenberg and give more flexibility on how images are inserted into the post.

I'm not sure if `edit-post` is the best choice here, but I just wanted to share the idea.

Usage:

```js
const replaceMediaUpload = () => MediaUpload;

addFilter(
	'blocks.MediaUpload',
	'core/edit-post/blocks/media-upload/replaceMediaUpload',
	replaceMediaUpload
);
```

By default `MediaUpload` renders a button which does nothing. The code above replaces this implementation with what we had before.

Let me know what do you think about this idea. If we agree on some variation of this approach I would have to extend docs with whatever is implemented in here.

## How Has This Been Tested?
Manually tested to confirm that there are no regressions. It should be still possible

Run `npm test` to make sure all tests still pass.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: blocks/media-upload/index.js [hunk 1]
```diff
@@ -1,156 +1,13 @@
...
-		return this.props.render( { open: this.openModal } );
-	}
-}
+/**
+ * WordPress dependencies
+ */
+import { withFilters } from '@wordpress/components';
 
-export default MediaUpload;
+const MediaUpload = ( { render } ) => render( { open: noop } );
```
Comment (gziolo) [inline blocks/media-upload/index.js]: We could also not render at all by default.

**Comment by youknowriad** [comment]
I think the best place for this component is in `wp.media` itself :) but probably not actionable at the moment.

---

That said, I'd like as part of this change to refactor slightly the blocks depending on this component to avoid rendering the empty buttons at all when this component is not available. Which would make the `editor` block work (maybe not optimally) but they will work just missing so UX enhancements added by the `MediaUpload` component. Does this make sense?

**Comment by gziolo** [comment]
> That said, I'd like as part of this change to refactor slightly the blocks depending on this component to avoid rendering the empty buttons at all when this component is not available. Which would make the `editor` block work (maybe not optimally) but they will work just missing so UX enhancements added by the `MediaUpload` component. Does this make sense?

I'm not quite sure I know what kind of refactoring for the blocks you mean. `<MediaUpload />` uses render props pattern, to provide custom UI depending on the context, e.g.:
```js
<MediaUpload
	gallery={ multiple }
	multiple={ multiple }
	onSelect={ onSelectImage }
	type="image"
	render={ ( { open } ) => (
		<Button isLarge onClick={ open }>
			{ __( 'Add from Media Library' ) }
		</Button>
	) }
/>
```

**Comment by gziolo** [comment]
Did you mean to conditionally display this button as follows:
```js
hasBlockSupport( settings, 'media' ) && <MediaUpload ... />
```

**Comment by youknowriad** [comment]
> Did you mean to conditionally display this button as follows:

I mean consider returning `null` from the `MediaUpload` button when no hook for this component is registered to avoid rendering non-functional buttons. 

But at the same time, make sure the block is still usable without these buttons. Make them an enhancement instead of a requirement to be able to use `editor` as a generic module without the need for media hooks.

**Comment by gziolo** [comment]
> I mean consider returning null from the MediaUpload button when no hook for this component is registered to avoid rendering non-functional buttons.

Right, I thought about that and even left a comment. However, it might not play well with toolbars when there are not other items:
```js
const MediaUpload = ( ) => null;
```
It needs some testing.

**Comment by youknowriad** [comment]
After some tests, I think we need to do this to ship this PR:

 - Return `null` by default
 - Add an "Upload" button to the audio and video blocks (similar to the image and gallery blocks)

These changes will make our blocks reusable even without the hook 👍

**Comment by gziolo** [comment]
> Add an "Upload" button to the audio and video blocks (similar to the image and gallery blocks)

This is a really good idea. It should unify how all those media related blocks behave. I will open a new PR to tackle it first.

**Comment by gziolo** [comment]
> Return `null` by default

Updated as discussed above :)

**Review by youknowriad — APPROVED** [review]
This is looking good to me 👍, Let's see what the others think @mtias @aduth

**Comment by gziolo** [comment]
I opened #5431 to add the `Upload` button for the audio and video blocks.

**Comment by gziolo** [comment]
> Add an "Upload" button to the audio and video blocks (similar to the image and gallery blocks)

This one is resolved with #5431. This PR should be good to go if we agree on the proposed solution.

**Comment by gziolo** [comment]
@jorgefilipecosta, I rebased this PR with your latest changes. Can you confirm that uploading images with a caption still work as expected? I tested with your test image:
36873864-bfdc6802-1da1-11e8-8071-311ce59494a3
It seems to work properly. I had to introduce a new function exposed from the `blocks` module to make sure it doesn't pick libraries that are already bundled with `blocks`.

File: edit-post/hooks/blocks/media-upload/index.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+	openModal() {
+		this.frame.open();
+	}
+
+	processMediaCaption( mediaObject ) {
+		return ! mediaObject.caption ?
+			mediaObject :
+			{ ...mediaObject, caption: parseWithAttributeSchema( mediaObject.caption, {
+				source: 'children',
```
Comment (jorgefilipecosta) [inline edit-post/hooks/blocks/media-upload/index.js]: Minor: edit post is showing a deep knowledge of how blocks work by using "source: 'children'",. I understand the reason for this but I think we can also create a function in parse called "parseHtml" or something similar that just receives a string and returns the parse result (calling parseWithAttributeSchema). So each time we want to pass html we just call it, it abstracts a little bit more how parsing works.

**Review by jorgefilipecosta — APPROVED** [review]
Nice work here, In my tests things worked fine 👍

File: edit-post/hooks/blocks/media-upload/index.js (hunk 2 above)
Comment (gziolo) [reply edit-post/hooks/blocks/media-upload/index.js]: I'm not sure what would be the best way to go. This one is very flexible and built on top of `hpq`. I didn't want to introduce too granular function which will be used only in one place. This is definitely something to look at and a subject to further refinement.

**Comment by gziolo** [comment]
@jorgefilipecosta thanks for performing a sanity check 👍

**Pull request merged**